staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go

```diff
@@ -489,214 +582,201 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+		"numChangedMatchedSlices", len(changedDesiredSlices),
+		"numNewSlices", numNewSlices,
+	)
+
+	bumpedGeneration := false
+	switch {
+	case pool.Generation > generation:
+		// Bump up the generation if the driver asked for it, or
+		// start with a non-zero generation.
+		generation = pool.Generation
+		bumpedGeneration = true
+		logger.V(5).Info("Bumped generation to driver-provided generation", "generation", generation)
+	case numNewSlices == 0 && len(changedDesiredSlices) <= 1:
+		logger.V(5).Info("Kept generation because at most one update API call is necessary", "generation", generation)
+	default:
+		generation++
+		bumpedGeneration = true
+		logger.V(5).Info("Bumped generation by one", "generation", generation)
+	}
+	desiredPool.Generation = generation
+
+	// First delete obsolete slices. If the desired slices are faulty, then it's still better to
+	// remove devices that the driver no longer has, even if we cannot publish the new ones.
+	if err := c.removeSlices(ctx, obsoleteSlices); err != nil {
+		return fmt.Errorf("remove slices: %w", err)
+	}
+
+	// Update existing slices.
+	for i, currentSlice := range currentSliceForDesiredSlice {
+		if !changedDesiredSlices.Has(i) && !bumpedGeneration {
+			continue
 		}
-		logger.V(5).Info("Completed comparison",
-			"numObsolete", len(obsoleteSlices),
-			"numMatchedSlices", len(currentSliceForDesiredSlice),
-			"numChangedMatchedSlices", len(changedDesiredSlices),
-			"numNewSlices", numNewSlices,
-		)
-
-		bumpedGeneration := false
-		switch {
-		case pool.Generation > generation:
-			// Bump up the generation if the driver asked for it, or
-			// start with a non-zero generation.
-			generation = pool.Generation
-			bumpedGeneration = true
-			logger.V(5).Info("Bumped generation to driver-provided generation", "generation", generation)
-		case numNewSlices == 0 && len(changedDesiredSlices) <= 1:
-			logger.V(5).Info("Kept generation because at most one update API call is necessary", "generation", generation)
-		default:
-			generation++
-			bumpedGeneration = true
-			logger.V(5).Info("Bumped generation by one", "generation", generation)
+		slice := currentSlice.DeepCopy()
+		slice.Spec.Pool = desiredPool
+		// No need to set the node name. If it was different, we wouldn't
+		// have listed the existing slice.
+		//
+		// When adding new fields here, then also extend sliceStored.
+		slice.Spec.NodeSelector = pool.NodeSelector
+		slice.Spec.AllNodes = desiredAllNodes
+		slice.Spec.SharedCounters = pool.Slices[i].SharedCounters
+		slice.Spec.PerDeviceNodeSelection = pool.Slices[i].PerDeviceNodeSelection
+		// Preserve TimeAdded from existing device, if there is a matching device and taint.
+		slice.Spec.Devices = copyTaintTimeAdded(slice.Spec.Devices, pool.Slices[i].Devices)
+
+		logger.V(5).Info("Updating existing resource slice", "slice", klog.KObj(slice))
+		actualSlice, err := c.kubeClient.ResourceV1beta1().ResourceSlices().Update(ctx, slice, metav1.UpdateOptions{})
```

Comment:
So here we actually do the update and then decide if fields were dropped. This means that the ResourceSlices with the missing fields are in the APIServer and will be used by the scheduler with potentially incorrect results. I remember we discussed trying to use dry-run to detect this situation without actually making the incorrect ResourceSlices live in the cluster. Did we try that out?

Reply:
No, I haven't. @johnbelamaric convinced me that plan A should be that admins downgrade drivers before downgrading the cluster. This check is relatively cheap and doesn't add much complexity, but a dry-run isn't, therefore I decided against doing it.